src/meta/src/stream/scale.rs

```diff
@@ -1474,3 +1489,255 @@ where
+            }
+
+            actor_status.extend(table_fragments.actor_status.clone());
+        }
+
+        for (
+            fragment_id,
+            WorkerChanges {
+                include_worker_ids,
+                exclude_worker_ids,
+            },
+        ) in fragment_worker_changes
+        {
+            let fragment = match fragment_map.get(&fragment_id).cloned() {
+                None => bail!("Fragment id {} not found", fragment_id),
+                Some(fragment) => fragment,
+            };
+
+            let include_worker_ids: HashSet<_> = include_worker_ids.into_iter().collect();
+            let exclude_worker_ids: HashSet<_> = exclude_worker_ids.into_iter().collect();
+
+            let intersection_ids = include_worker_ids
+                .intersection(&exclude_worker_ids)
+                .collect_vec();
+
+            if !intersection_ids.is_empty() {
+                bail!(
+                    "Include worker ids {:?} and exclude worker ids {:?} have intersection {:?}",
+                    include_worker_ids,
+                    exclude_worker_ids,
+                    intersection_ids
+                );
+            }
+
+            for worker_id in include_worker_ids.iter().chain(exclude_worker_ids.iter()) {
+                if !worker_parallel_units.contains_key(worker_id) {
+                    bail!("Worker id {} not found", worker_id);
+                }
+            }
+
+            let fragment_parallel_unit_ids: HashSet<_> = fragment
+                .actors
+                .iter()
+                .map(|actor| {
+                    actor_status
+                        .get(&actor.actor_id)
+                        .and_then(|status| status.parallel_unit.clone())
+                        .unwrap()
+                        .id as ParallelUnitId
+                })
+                .collect();
+
+            match fragment.get_distribution_type().unwrap() {
+                FragmentDistributionType::Unspecified => unreachable!(),
+                FragmentDistributionType::Single => {
+                    assert_eq!(fragment_parallel_unit_ids.len(), 1);
+
+                    let single_parallel_unit_id =
+                        fragment_parallel_unit_ids.iter().exactly_one().unwrap();
+
+                    let target_parallel_unit_ids: HashSet<_> = worker_parallel_units
+                        .keys()
+                        .chain(include_worker_ids.iter())
+                        .filter(|id| !exclude_worker_ids.contains(*id))
+                        .flat_map(|id| worker_parallel_units.get(id).cloned().unwrap())
+                        .collect();
+
+                    if !target_parallel_unit_ids.contains(single_parallel_unit_id) {
+                        // todo: choose a target parallel unit by single actor count
+                        let chosen_target_parallel_unit_id =
+                            target_parallel_unit_ids.iter().next().unwrap();
```

Comment:
Previously, the consideration here was to decide where to schedule through a deterministic algorithm like the number of single fragment's actors. However, it seems that it might be better to determine this through actual operation, so a random method was chosen. I'll try using a deterministic method instead and add some comments.